### PR TITLE
Use createElement instead of HTML generation

### PR DESCRIPTION
**Pull request by spicyj**
Behind a feature flag. This is a relatively simple change; adopting this strategy universally would mean that we could clean up a lot of code but this doesn't attempt to restructure more than necessary.

### CODE REVIEW

**Comment by zpao** [comment]
oh 💩!

**Comment by spicyj** [comment]
If you turn it on, most of the tests fail because the nodes are from the wrong document and jsdom throws some exception. Not sure how best to thread the document through… for the mount case we could attach it to the transaction but that seems wrong for the update case when we update multiple roots. Context? :D

**Comment by spicyj** [comment]
But Facebook (the website) seems to work fine with it on so.

**Comment by syranide** [comment]
FYI #3146

**Comment by yaycmyk** [comment]
:smile:

**Comment by sebmarkbage** [comment]
What's the goal of doing this incrementally instead of creating two big forks?

The forks could compete on their own merit instead of trying to please each other.

**Comment by spicyj** [comment]
Well, this was less typing and makes it easier to run them side by side for an A/B test or similar. I figured if createElement shows comparable or better metrics then we can rip out the HTML code entirely and not worry about it (though I guess we might still need to keep it for server rendering and rehydration?).

**Comment by syranide** [comment]
> I figured if createElement shows comparable or better metrics then we can rip out the HTML code entirely and not worry about it

@spicyj @petehunt did some tests on this a long time back. IIRC it performed really really well when used together with `cloneNode` to duplicate common structures (say, components with many similar hierarchies, which is really common).

File: src/renderers/dom/client/ReactMount.js [hunk 1]
```diff
@@ -294,7 +313,9 @@ function batchedMountComponentIntoNode(
   shouldReuseMarkup,
   context
 ) {
-  var transaction = ReactUpdates.ReactReconcileTransaction.getPooled();
+  var transaction = ReactUpdates.ReactReconcileTransaction.getPooled(
+    /* forceHTML */ shouldReuseMarkup
```
Comment (zpao) [inline src/renderers/dom/client/ReactMount.js]: I think in the cases where we've labeled args like this, we've done it after, not before (though I could be wrong, not sure if we actually have any in React).

File: src/renderers/dom/client/wrappers/__tests__/ReactDOMTextarea-test.js [hunk 2]
```diff
@@ -62,7 +62,7 @@ describe('ReactDOMTextarea', function() {
   });
 
   it('should display "false" for `defaultValue` of `false`', function() {
-    var stub = <textarea type="text" defaultValue={false} />;
+    var stub = <textarea defaultValue={false} />;
```
Comment (zpao) [inline src/renderers/dom/client/wrappers/__tests__/ReactDOMTextarea-test.js]: Nice cleanup…

File: src/renderers/dom/client/ReactMount.js (hunk 1 above)
Comment (spicyj) [inline src/renderers/dom/client/ReactMount.js]: If the language actually had labeled args, the labels would certainly go before…

File: src/renderers/dom/client/wrappers/__tests__/ReactDOMTextarea-test.js (hunk 2 above)
Comment (spicyj) [inline src/renderers/dom/client/wrappers/__tests__/ReactDOMTextarea-test.js]: (`.type` is always `'textarea'`; setting it throws.)

File: src/renderers/dom/shared/ReactDOMComponent.js [hunk 3]
```diff
@@ -1040,7 +1076,7 @@ assign(
 
 ReactDOMComponent.injection = {
   injectIDOperations: function(IDOperations) {
-    ReactDOMComponent.BackendIDOperations = BackendIDOperations = IDOperations;
+    ReactDOMComponent.BackendIDOperations = IDOperations;
```
Comment (zpao) [inline src/renderers/dom/shared/ReactDOMComponent.js]: ReactDOMTextArea is the only thing left using `ReactDOMComponent.BackendIDOperations` - should we update that while we're here and remove the injection?

Comment (spicyj) [inline src/renderers/dom/shared/ReactDOMComponent.js]: Guess so.

File: src/renderers/dom/client/ReactDOMIDOperations.js [hunk 4]
```diff
@@ -91,14 +90,14 @@ var ReactDOMIDOperations = {
    * @param {string} name A property name to remove, see `DOMProperty`.
    * @internal
    */
-  deletePropertyByID: function(id, name, value) {
+  deletePropertyByID: function(id, name) {
```
Comment (zpao) [inline src/renderers/dom/client/ReactDOMIDOperations.js]: I don't think this is actually used…

Comment (zpao) [inline src/renderers/dom/client/ReactDOMIDOperations.js]: Actually, I wonder if that's true for a number of things in here now that it's not being injected.

Comment (spicyj) [inline src/renderers/dom/client/ReactDOMIDOperations.js]: shh it could have been! but yeah, I'll remove the ones that aren't.

Comment (zpao) [inline src/renderers/dom/client/ReactDOMIDOperations.js]: Hey, *you* asked *me* for this review, just doing my job. You didn't think I'd just stamp this did you? 😝

Now if you hadn't changed this method at all, it probably would have slipped past.

**Comment by zpao** [comment]
Let's give it a go.

**Comment by spicyj** [comment]
\o/

**Pull request merged**